docs/developer/logging-guidelines.md

```diff
@@ -0,0 +1,407 @@
+
+   ```go
+    logger.Info("Got response",
+        "status", "ok")
+   ```
+
+3. For readability, put each key-value pair on a separate line, as in the previous example.
+4. Use camelCase for key names.
+5. When dealing with a Kubernetes resource, use `resource` as a key:
+
+   ```go
+    logger.Info("processed resource",
+        "resource", hr)
+   ```
+
+6. Don't put structs as values expect for Kubernetes resources, to prevent messy and potentially unbounded output.
+
+### Log Formatting
+
+By default, log messages are formatted as JSON, so that external systems can easily parse them.
+
+For the developer, log messages are formatted as text strings (except key/values), for human-friendly parsing. For
+example:
+
+```text
+2023-07-21T12:41:37.640-0400    INFO    processed resource      {"resource": {"apiVersion": "gateway.networking.k8s.io/v1beta1", "kind": "HTTPRoute", "namespace": "test", "name": "hr-1"}}
+```
+
+The formatting is controlled during the logger initialization.
+
+### When to Log
+
+Below are the cases for when to log a message:
+
+- Lifecycle events
+  - The control plane starts or shuts down
+  - Lifecycle of the components.
+- Handling a transaction/request
+  - The main transaction in the control plane is processing a batch of events related to Kubernetes resources taken
+      from the main event loop.
+  - Additional transactions could be out of the event loop requests to auxiliary control plane endpoints (APIs).
+- Interaction with external components. When the control plane interacts with an external component, log the stages of
+  that interaction as well as the result of it.
+
+For those cases:
+
+- Log an error only if the code doesn't handle it (performs some action). If it is handled, but it is still important to
+  record the fact of that
+  error, log it as an info message.
+- Consider including performance-related information. For example, the duration of an operation.
+- Log extra details (or steps) to facilitate troubleshooting under the `debug` level.
+
+> It is important to not overload the `info` level with too many details and messages.
+
+### What not to Log
+
+- *Sensitive information*, including the contents of the secrets, passwords, etc.
+- *Returned errors*. If a function returns an error, do not log it inside that function, because it will be logged or
+  handled
+  up in the stack.
+- *Critical errors*, for which we [panic](go-style-guide.md#use-panics-for-unrecoverable-errors-or-programming-errors).
+
+### Performance
+
+Logging too frequently can lead to poor performance, because of:
+
+- Too many I/O operations.
+- Too many allocations related to preparing the log message.
+
+If we diligently follow the guidelines from this document, the `error` and `info` log levels will not have many messages
+(many thousands per second). At the same time, such a case might still be possible, for example, in the case of a bug.
+To prevent that, the default settings in the controller-runtime helpers for zap
```

Comment:
So we currently have sampling enabled? Is that what we want?

Reply:
it is something that will prevent logging overflow and, as a result, possible performance degradation, in case there is a bug. For example, there some error that NKG keeps constantly retrying in a loop.
so I think it is a reasonable safeguard to have that enabled.
At the same time,  my concern is that we don't directly control sampling -- it is configured with zap helpers.

Reply:
I'm not familiar with zap helpers, but it is possible to disable them and directly control sampling? Or maybe write our own helper?

Reply:
yeah. we can write out own helpers.

those helpers are here https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/log/zap/zap.go

we use them when we create the logger:
```
logger := zap.New()
```

the logger is still customizable via those helpers via Options -- https://github.com/kubernetes-sigs/controller-runtime/blob/21779fbe6419e03ae1cd86a69b35d2725c6f7558/pkg/log/zap/zap.go#L132-L167

also, it includes KubeAwareEncoder that allows us to put a whole kubernetes resource as an object -- https://github.com/kubernetes-sigs/controller-runtime/blob/21779fbe6419e03ae1cd86a69b35d2725c6f7558/pkg/log/zap/zap.go#L240 

however, sampling is always on (except for some cases) -- https://github.com/kubernetes-sigs/controller-runtime/blob/21779fbe6419e03ae1cd86a69b35d2725c6f7558/pkg/log/zap/zap.go#L203-L208

Reply:
Thanks for all the info. Is it possible to disable the sampling? Or are we stuck with it if we use zap?

Reply:
> Thanks for all the info. Is it possible to disable the sampling? Or are we stuck with it if we use zap?

we can disable it.  in this case, we will create Zap without using those helpers -- we will create it directly.

shall I add an issue to the list for that? and update the guide with regards to sampling

Reply:
I'm not sure what the best choice is here. It's good to know that we can disable sampling, but we don't have a justification for doing so besides that the underlying implementation may change from under us.

I'm leaning toward leaving it enabled for now. If we run into issues with it in the future at least we know we can disable it.